src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php

```diff
@@ -0,0 +1,133 @@
+                $node instanceof EnumNode => $node->getPermissibleValues('|'),
+                $node instanceof ScalarNode => 'string|int|float|bool|null',
+                $node instanceof VariableNode => 'mixed',
+            };
+        }
+
+        if ($node instanceof PrototypedArrayNode) {
+            $isHashmap = (bool) $node->getKeyAttribute();
+
+            return 'array<'.($isHashmap ? 'string, ' : '').self::doGeneratePhpDoc($node->getPrototype(), 1 + $nestingLevel).'>';
+        }
+
+        if (!($children = $node->getChildren()) && !$node->getParent() instanceof PrototypedArrayNode) {
+            return 'array<mixed>';
+        }
+
+        $arrayShape = \sprintf("array{%s\n", self::generateInlinePhpDocForNode($node));
+
+        foreach ($children as $child) {
+            $arrayShape .= str_repeat('    ', $nestingLevel).self::dumpNodeKey($child).': ';
+
+            if ($child instanceof PrototypedArrayNode) {
+                $isHashmap = (bool) $child->getKeyAttribute();
+
+                $arrayShape .= 'array<'.($isHashmap ? 'string, ' : '').self::doGeneratePhpDoc($child->getPrototype(), 1 + $nestingLevel).'>';
+            } else {
+                $arrayShape .= self::doGeneratePhpDoc($child, 1 + $nestingLevel);
+            }
+
+            $arrayShape .= \sprintf(",%s\n", !$child instanceof ArrayNode ? self::generateInlinePhpDocForNode($child) : '');
+        }
+
+        if ($node->shouldIgnoreExtraKeys()) {
+            $arrayShape .= str_repeat('    ', $nestingLevel)."...<mixed>\n";
+        }
+
+        return $arrayShape.str_repeat('    ', $nestingLevel - 1).'}';
+    }
+
+    private static function dumpNodeKey(NodeInterface $node): string
+    {
+        $name = $node->getName();
+        $quoted = str_starts_with($name, '@')
+            || \in_array(strtolower($name), ['int', 'float', 'bool', 'null', 'scalar'], true)
+            || strpbrk($name, '\'"');
+
+        if ($quoted) {
+            $name = "'".addslashes($name)."'";
+        }
+
+        return $name.($node->isRequired() ? '' : '?');
+    }
+
+    private static function handleNumericNode(NumericNode $node): string
+    {
+        $min = $node->getMin() ?? 'min';
+        $max = $node->getMax() ?? 'max';
+
+        if ($node instanceof IntegerNode) {
+            return \sprintf('int<%s, %s>', $min, $max);
+        }
+        if ($node instanceof FloatNode) {
+            return 'float';
+        }
+
+        return \sprintf('int<%s, %s>|float', $min, $max);
+    }
+
+    private static function generateInlinePhpDocForNode(BaseNode $node): string
+    {
+        $comment = '';
+        if ($node->hasDefaultValue() || $node->getInfo() || $node->isDeprecated()) {
+            if ($node->isDeprecated()) {
+                $comment .= 'Deprecated: '.$node->getDeprecation($node->getName(), $node->getPath())['message'].' ';
+            }
+
+            if ($info = $node->getInfo()) {
+                $comment .= $info.' ';
```

Comment:
shouldn't we have some kind of delimiter between the deprecation message and the info (and the default) for cases using multiple sources of comment ?